saleor/dashboard/product/forms.py

```diff
@@ -123,19 +124,34 @@ def clean(self):
                 'and its variant.')
             self.add_error('variant_attributes', msg)
 
-        if self.instance.pk:
-            variants_changed = (
-                self.fields['has_variants'].initial != has_variants)
-            if variants_changed:
-                query = self.instance.products.all()
-                query = query.annotate(variants_counter=Count('variants'))
-                query = query.filter(variants_counter__gt=1)
-                if query.exists():
-                    msg = pgettext_lazy(
-                        'Product type form error',
-                        'Some products of this type have more than '
-                        'one variant.')
-                    self.add_error('has_variants', msg)
+        if not self.instance.pk:
+            return data
+
+        variants_changed = (
+            self.fields['has_variants'].initial != has_variants)
+        if variants_changed:
+            query = self.instance.products.all()
+            query = query.annotate(variants_counter=Count('variants'))
+            query = query.filter(variants_counter__gt=1)
+            if query.exists():
+                msg = pgettext_lazy(
+                    'Product type form error',
+                    'Some products of this type have more than '
+                    'one variant.')
+                self.add_error('has_variants', msg)
+
+        # Some variant attributes could be removed so name should be updated
+        # accordingly
+        initial_attributes = set(list(self.instance.variant_attributes.all()))
```

Comment:
This `clean` is huuuge. Maybe we could benefit from having 2-3 helper functions that do these particular pieces of logic?